### PR TITLE
KAFKA-15625: Do not flush global state store at each commit

### PR DESCRIPTION
**Pull request by AyoubOm**
Global state stores are currently flushed at each commit, which may impact performance, especially for EOS (commit each 200ms).
The goal of this improvement is to flush global state stores only when the delta between the current offset and the last checkpointed offset exceeds a threshold.
This is the same logic we apply on local state store, with a threshold of 10.000 records.
The implementation replaces the flush based on time to a flush based on the delta. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateUpdateTask.java [hunk 1]
```diff
@@ -150,5 +150,11 @@ private void initTopology() {
         }
     }
 
+    @Override
+    public void maybeCheckpoint() {
+        if (StateManagerUtil.checkpointNeeded(false, stateMgr.changelogOffsets(), offsets)) {
```
Comment (jeffkbkim) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateUpdateTask.java]: [AbstractTask#maybeCheckpoint()](https://github.com/apache/kafka/blob/a1f3c6d16061566a4f53c72a95e2679b8ee229e0/streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java#L97) passes in `stateMgr.changelogOffsets` as the new offset snapshot in `StateManagerUtil#checkpointNeeded`. Why is the behavior reversed here?

is it because `offsets` is the object that is updated on every `update()`? what about `ProcessorStateManager#changelogOffsets`, would that not have the latest?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java [hunk 2]
```diff
@@ -259,19 +251,14 @@ void initialize() {
...
             final ConsumerRecords<byte[], byte[]> received = globalConsumer.poll(pollTime);
             for (final ConsumerRecord<byte[], byte[]> record : received) {
                 stateMaintainer.update(record);
             }
-            final long now = time.milliseconds();
-            if (now - flushInterval >= lastFlush) {
-                stateMaintainer.flushState();
-                lastFlush = now;
-            }
+            stateMaintainer.maybeCheckpoint();
```
Comment (jeffkbkim) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java:261]: should we still keep an interval-based flush or is the delta check sufficient?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateConsumerTest.java [hunk 3]
```diff
@@ -161,6 +140,10 @@ public void update(final ConsumerRecord<byte[], byte[]> record) {
             updatedPartitions.put(tp, updatedPartitions.get(tp) + 1);
         }
 
+        @Override
+        public void maybeCheckpoint() {
+            flushState();
+        }
     }
 
 }
```
Comment (jeffkbkim) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateConsumerTest.java]: nit: can we add a newline?

**Review by jeffkbkim — COMMENTED** [review]
thanks for the PR. left some comments

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateUpdateTask.java (hunk 1 above)
Comment (AyoubOm) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateUpdateTask.java]: Hi Jeff
For GlobalStateManager, the method _changelogOffsets()_ returns the last checkpointed offset, it is updated only on checkpoint, whereas the field _offsets_ of GlobalStateUpdateTask is as you said updated for each record, hence the order here.
For ProcessorStateManager the method computes the offset from StateStoreMetadata which is insync with the last offset of the changelog topic

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java (hunk 2 above)
Comment (AyoubOm) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java:261]: For me, the idea is to only flush&checkpoint based on the delta of records, as we do for task state stores

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateConsumerTest.java (hunk 3 above)
Comment (AyoubOm) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateConsumerTest.java]: done

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java [hunk 4]
```diff
@@ -217,21 +215,53 @@ public void shouldFlushStateManagerWithOffsets() {
...
+
+        assertEquals(offsets, stateMgr.changelogOffsets());
+        assertFalse(stateMgr.flushed);
+        assertFalse(stateMgr.checkpointWritten);
+    }
+
+    @Test
+    public void shouldCheckpointIfReceivedEnoughRecords() {
+        final Map<TopicPartition, Long> expectedOffsets = new HashMap<>();
+        expectedOffsets.put(t1, 10051L); // t1 advanced with 10.001 records
```
Comment (jeffkbkim) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java]: nit: should this be 10,001 or 10001?

**Review by jeffkbkim — APPROVED** [review]
left a minor comment. Otherwise, LGTM

test failures look unrelated

**Comment by jeffkbkim** [comment]
@cadonna can you take a look?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java (hunk 4 above)
Comment (AyoubOm) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java]: fixed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java (hunk 2 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java:261]: I think, we should still keep an interval-based flush. However, we should not flush if the offset delta is not larger than `OFFSET_DELTA_THRESHOLD_FOR_CHECKPOINT`. If we remove the interval-based flush that would be quite a change in behavior. Users that set the commit interval to large values would not expect to see flushes even though the offset delta is exceeded.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java [hunk 5]
```diff
@@ -217,21 +215,53 @@ public void shouldFlushStateManagerWithOffsets() {
...
+    }
+
+    @Test
+    public void shouldCheckpointIfReceivedEnoughRecords() {
+        final Map<TopicPartition, Long> expectedOffsets = new HashMap<>();
+        expectedOffsets.put(t1, 10051L); // t1 advanced with 10001 records
+        expectedOffsets.put(t2, 100L);
+
+        globalStateTask.initialize();
+        globalStateTask.update(record(topic1, 1, 10050L, "foo".getBytes(), "foo".getBytes()));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java]: Could you please additionally also do an update here just before (i.e. at 10049) the threshold and verify that no flush was performed and no checkpoint was written?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:1]: Why were these changes needed? You did not change anything in `StreamTask`.

**Review by cadonna — COMMENTED** [review]
Thanks for the PR @AyoubOm !

Here my feedback!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java (hunk 2 above)
Comment (AyoubOm) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java:261]: Thanks @cadonna for your comment. Yes, I agree with your point.
However, wouldn't that create a difference in behavior between global and local state stores ? I saw that flushing local state stores is based only on the delta. Please correct me if I am wrong

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java (hunk 5 above)
Comment (AyoubOm) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java]: Actually, that is the goal of the test above this one. I will replace 9000 by 10049

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (AyoubOm) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:1]: Yes, that was more of an improvement of the test: we checkpoint both on restoration and on postCommit.
We were only testing that checkpoint is called at least once.

**Comment by AyoubOm** [comment]
> Thanks for the PR @AyoubOm !
> 
> Here my feedback!

Thanks @cadonna for your review ! For your first comment please check my reply to it above.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java (hunk 5 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java]: I know, I wanted to have a verification for the boundary. Could you please re-add the update with offset 9000 and additionally do an update with offset 10049. Then we also test the case with multiple updates before the checkpoint.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java (hunk 2 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java:261]: No, checkpointing the local store is not only based on the delta. Method `maybeCheckpoint()` is called after a commit and a commit is done in regular time-intervals.
https://github.com/apache/kafka/blob/d144b7ee387308a59e52cbdabc7b66dd3b2926cc/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L513

**Review by cadonna — COMMENTED** [review]
@AyoubOm Thanks for the updates!

Here my replies.

**Comment by AyoubOm** [comment]
> @AyoubOm Thanks for the updates!
> 
> Here my replies.

Thanks @cadonna, I made updates accordingly

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 6]
```diff
@@ -1664,18 +1664,17 @@ public void shouldNotCheckpointOffsetsAgainOnCommitIfSnapshotNotChangedMuch() {
...
             .andReturn(singletonMap(changelogPartition, 10L))
             .andReturn(singletonMap(changelogPartition, 20L));
-        EasyMock.expectLastCall();
         EasyMock.replay(stateManager, recordCollector);
 
         task = createStatefulTask(createConfig("100"), true);
 
         task.initializeIfNeeded();
-        task.completeRestoration(noOpResetter -> { });
+        task.completeRestoration(noOpResetter -> { }); // should checkpoint
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: These changes do not seem necessary. Could you please revert them?

**Review by cadonna — COMMENTED** [review]
Thanks for the updates @AyoubOm !

I have just one minor comment. When that comment will be addressed, I am going to merge the PR.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 6 above)
Comment (AyoubOm) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Done. Maybe I will create a minor separate PR for that

**Review by cadonna — APPROVED** [review]
Thanks @AyoubOm !

LGTM!

**Comment by cadonna** [comment]
Failed tests are unrelated.
